### PR TITLE
Fix initial population of FormSets from an InProgressSubmission

### PR DESCRIPTION
**Pull request by yenaing-oo**
Closes https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/issues/241

### CODE REVIEW

File: bagitobjecttransfer/recordtransfer/enums.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    RIGHTS = "rights"
+    OTHER_IDENTIFIERS = "otheridentifiers"
+    GROUP_TRANSFER = "grouptransfer"
+    UPLOAD_FILES = "uploadfiles"
+    FINAL_NOTES = "finalnotes"
+
+    @classmethod
+    def get_formset_steps(cls) -> list["TransferStep"]:
+        """Return the steps of the multi-stage form which use formsets."""
+        return [TransferStep.RIGHTS, TransferStep.OTHER_IDENTIFIERS]
```
Comment (danloveg) [inline bagitobjecttransfer/recordtransfer/enums.py:23]: I recommend removing this function (see my later comments for why)

**Review by danloveg — CHANGES_REQUESTED** [review]
Looks like it's mostly working. I'm still seeing an extra empty form being added on the rights page when resuming a transfer, and I noted some areas where we can simplify the code in the transfer form wizard.

File: bagitobjecttransfer/recordtransfer/views.py [hunk 2]
```diff
@@ -527,8 +585,20 @@ def get_context_data(self, form, **kwargs):
...
         return context
 
+    @property
+    def num_extra_forms(self) -> int:
+        """Compute the number of extra forms to generate if current step uses a formset."""
+        num_extra_forms = 1
+        if self.current_step in [TransferStep.RIGHTS, TransferStep.OTHER_IDENTIFIERS]:
+            num_forms = len(self.get_form_initial(self.current_step.value))
+            num_extra_forms = 0 if num_forms > 0 else 1
+        return num_extra_forms
```
[the remark this answers is not in the document]
Comment (yenaing-oo) [reply bagitobjecttransfer/recordtransfer/views.py:600]: I double-checked on my end by rebuilding the container and calling `collectstatic --clear --noinput`. It appears to work fine for me when I save and resume with only one form filled on the Rights step:

image

Did you call `collectstatic` before testing by any chance?

File: bagitobjecttransfer/recordtransfer/views.py [hunk 3]
```diff
@@ -361,20 +386,27 @@ def get(self, request, *args, **kwargs):
...
 
         past_data = self.storage.data
-        current_data = TransferFormWizard.format_step_data(save_form_step, request.POST)
+        current_data = None
+        if self.current_step in TransferStep.get_formset_steps():
+            current_data = TransferFormWizard.format_formset_step_data(
+                self.current_step, request.POST
+            )
+        else:
+            current_data = TransferFormWizard.format_step_data(self.current_step, request.POST)
```
[the remark this answers is not in the document]
Comment (yenaing-oo) [reply bagitobjecttransfer/recordtransfer/views.py]: Resolved by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/243/commits/917757cfee3f8423c9d3047f56c81233dfbab1c3

File: bagitobjecttransfer/recordtransfer/enums.py (hunk 1 above)
Comment (yenaing-oo) [reply bagitobjecttransfer/recordtransfer/enums.py]: Resolved by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/243/commits/917757cfee3f8423c9d3047f56c81233dfbab1c3

File: bagitobjecttransfer/recordtransfer/views.py [hunk 4]
```diff
@@ -393,17 +425,42 @@ def load_transfer_data(self, transfer: InProgressSubmission) -> None:
...
+                index = int(parts[1])
+                field_name = parts[2]
+                if index not in indexed_data:
+                    indexed_data[index] = {}
+                indexed_data[index][field_name] = value
+
+        for index in sorted(indexed_data):
+            formatted_data.append(indexed_data[index])
+
+        return formatted_data
```
[the remark this answers is not in the document]
Comment (yenaing-oo) [reply bagitobjecttransfer/recordtransfer/views.py]: The new regex is very clean and accounts for both cases! Using the optional index group in the middle of the pattern was neat.

Resolved by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/243/commits/917757cfee3f8423c9d3047f56c81233dfbab1c3

File: bagitobjecttransfer/recordtransfer/views.py [hunk 5]
```diff
@@ -464,11 +520,14 @@ def get_form_initial(self, step):
         initial = self.initial_dict.get(step, {})
 
         if self.in_progress_submission and step == self.in_progress_submission.current_step:
-                data = pickle.loads(self.in_progress_submission.step_data)["current"]
+            data = pickle.loads(self.in_progress_submission.step_data)["current"]
+            if step in [s.value for s in TransferStep.get_formset_steps()]:
+                initial = data
+            else:
                 for k, v in data.items():
                     initial[k] = v
```
[the remark this answers is not in the document]
Comment (yenaing-oo) [reply bagitobjecttransfer/recordtransfer/views.py]: Resolved by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/243/commits/917757cfee3f8423c9d3047f56c81233dfbab1c3

File: bagitobjecttransfer/recordtransfer/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danloveg) [reply bagitobjecttransfer/recordtransfer/views.py:597]: That fixed the issue!

I keep getting tripped up by having to call `collectstatic` in development, I'm going to open a new issue to add it to the entrypoint script when running in development

File: bagitobjecttransfer/recordtransfer/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (danloveg) [reply bagitobjecttransfer/recordtransfer/views.py]: I find optional matching groups to be a super useful regex feature, glad you like it!

**Review by danloveg — APPROVED** [review]
Looks great. Just tested it out and everything is working as it should

**Pull request merged**